### DISCUSSION TITLE
Summon Boat spell, selection of the boat summoned when casting the "Summon Boat" spell

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#2588] [Ideas]
For the selection of the boat summoned when casting the "Summon Boat" spell, I suggest that it works this way:
- the free neutral boats (empty boats at the beginning of the scenario) must be summoned first according to their distance, the closest one first,

- a boat that has just been summoned is ignored for subsequent summonings as long as the hero is located in the same square where it was during the previous summonings,

- then are summoned the empty boats built in the player's castles in the reverse order of construction in time (the one built most recently first, then the one built just before, or the new one built just after the first summoned boat was built, etc.), 

- then if the hero casting the spell occupied a boat, that boat is summoned first, unless that boat is already located in a square adjacent to the hero casting the spell,

- then are summoned the empty boats belonging to the player which have been summoned but have not been embarked by a player's hero,

- then are summoned the empty boats freed by the player's other heroes in relation to time, the one freed longest ago first.

Note:
- When a hero summons a free boat (neutral at the beginning of the scenario), this boat is now considered as belonging to the player who leads the hero, and it cannot be summoned by a hero of another player.

- When a player builds a boat in one of his castles, this boat is considered as belonging to the player and cannot be summoned by a hero of another player.

- Similarly, when a hero boards a boat built by another player, or invoked or previously occupied by a hero of another player, that boat is now considered as belonging to the player and cannot be summoned by a hero of another player.

Precisions:
In Heroes 2, the Summon Boat spell does not create a boat.
In order to summon a boat, there must be a neutral boat on the map (empty boat at the beginning of the scenario), or the player must build a shipyard and then build a boat, or use another player's boat (the boat is then marked as his own, because the hero was the last occupant) and disembark from that boat.

**Edit.**

Related issue: https://github.com/ihhub/fheroes2/issues/2175.

**Comment by Branikolog** [comment]
I suggest summoning in more simple way:
Hero summons ALWAYS the nearest neutral or the boat, that was built or previously used by the player before. This will be more predictable, as player won't need to remember all those rules of summoning.

I prefer the majority of rules above working simultaneously.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> I suggest summoning in more simple way:
> Hero summons ALWAYS the nearest neutral or the boat, that was built or previously used by the player before. This will be more predictable, as player won't need to remember all those rules of summoning.
> 
> I prefer the majority of rules above working simultaneously.

This will be completely hidden to the player and he will not have to remember these rules.
These rules are made just so that the player doesn't have to worry about them.

All these rules work at the same time, it's just an enumeration of the order of priority in which the spell should summon the boats from the programming point of view.

Your sentence corresponds almost completely to what is enumerated. :)

> Hero summons ALWAYS the nearest neutral...

This is the first point.

> ...or the boat, that was built...

This is the third point.

> ...or previously used by the player before.

It's almost the fourth point, but you're talking about player instead of hero.

The second point is also important because it allows, for example, as in the original game, a hero to summon several boats before boarding one of them.
The other boats are destined for other of his heroes who have no possibility to summon a boat and who will board later, the next turn for example.

The fifth and sixth points are there to make sure that the boats summoned or last occupied by the other heroes of the player are summoned last, to avoid "stealing" the boats of the other heroes and who could therefore find themselves without a boat to board and without being able to summon one.

Now, I have also thought of another mode of operation of the "Summon Boat" spell which would be more practical by changing the mode of operation of the spell:

- If there is a free neutral boat (empty boat at the beginning of the scenario), the nearest free boat is summoned.
This boat is now considered as belonging to the player who leads the hero (so it's no longer a free boat).

- If there is no free boat, the spell is not cast immediately, the mouse cursor is changed, and the player has to select directly on the adventure map (with a left click) one of the boats he owns (marked as one of his own) for the spell to be cast.
The player should also be able to undo the spell without losing spell points (with a right click anywhere).

So, more efficient but with a little more interaction from the player. :)

**Edit.**

By activating the mouse wheel (when the spell is cast), the view of the adventure map is centered on a different boat belonging to the player for each click of the wheel.

On the minimap, the position of all the boats belonging to the player is indicated with icons that blink (when the spell is cast).

The player can click directly on the minimap (even when the spell is cast) to change the view of the adventure map.

**Reply by Branikolog** [reply]
> The fifth and sixth points are there to make sure that the boats summoned or last occupied by the other heroes of the player are summoned last, to avoid "stealing" the boats of the other heroes and who could therefore find themselves without a boat to board and without being able to summon one.

Nevertheless, in this case I would still prefer summoning here the nearest boat. Even if the hero unboarded nearby could be "robbed".) 

I'm still not sure, if I understand you correctly, but don't see the need in all that rules, I mean, l "the last boat used by the hero" or "boats that were built in reverse order"
My idea is to always summon the nearest boat. No matter neutral it, or was built in the nearest castle, or any player's hero disembarked nearby. Surely that should be done considering the way fheroes2 scanning the area. (If I'm not mistaken, in the original game for some cases priority scanning is often done clockwise)

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> > The fifth and sixth points are there to make sure that the boats summoned or last occupied by the other heroes of the player are summoned last, to avoid "stealing" the boats of the other heroes and who could therefore find themselves without a boat to board and without being able to summon one.
> 
> Nevertheless, in this case I would still prefer summoning here the nearest boat. Even if the hero unboarded nearby could be "robbed".)

Not for me, it's already one of the problems of the original game.
In the original game, very often it is precisely the boat that was not supposed to be summoned that it is summoned.
And that becomes a problem, because that boat belonged to another hero who was doing actions and who needed that boat afterwards.
The worst thing is when that happens and you realize that the boat is gone a few turns later, and the whole strategy is ruined.

No, this is a problem that needs to be fixed in fheroes2.

> I'm still not sure, if I understand you correctly, but don't see the need in all that rules, I mean, l "the last boat used by the hero"...

This is to avoid summoning a boat that has just been summoned by another hero ("stealing" his boat).
If the hero already occupied a boat before, it is "his boat" that is summoned.

> ...or "boats that were built in reverse order"

One of your heroes needs a boat.
You only have enough resources to build one boat.
A boat has already been built in one castle (or several in several castles), but it is destined for another hero.
You then build the necessary boat in a castle and thanks to the rule of the last boat built in a castle in the reverse order of construction in time, it is the last boat built that is summoned and not the previous one which is destined for another hero.

> My idea is to always summon the nearest boat. No matter neutral it, or was built in the nearest castle, or any player's hero disembarked nearby. Surely that should be done considering the way fheroes2 scanning the area.

Ah, always the closest boat regardless of its "state".

In this case we fall back into the current problem, where it is always the same boat that is summoned if the hero casts the spell several times in a row.
Because as soon as the hero casts the spell once, the boat becomes the closest one and it is always this boat that is summoned each time the hero casts the spell.
It is then no longer possible to summon several boats in all the squares adjacent to the hero.

I would prefer that the spell work as in my last proposal.
If there is no free boat on the map (empty boat at the beginning of the scenario), the player chooses the boat to summon from among those he owns.

It's a simple way to get the spell to achieve the desired result (cast the spell, choose the boat), and there's no need for all these rules anymore. :)

> (If I'm not mistaken, in the original game for some cases priority scanning is often done clockwise)

I don't know how it works in the original game.
Based on the coordinates of the boats? And with certain rules?

All I know is that it's not predictable and it's painful to use. :)

**Reply by Branikolog** [reply]
Dear @LeHerosInconnu, I understand the rules you want to apply. They help to automatically summon a boat with the best conditions for a player. 
(Except summoning neutral boats first. I think it's unfair, as hero , that summoned neutral boats while having his own being built or being owned before,  can rob other players. (For example, I own two boats in the beginning of the game. There's also two neutral boats somewhere on a map. I summon them from the far distant area. => now I own 4 boats. The next turn, you're trying to summon boats and it fails.)
But I'm so used to this "painful" spell, as it is.)) For me it's one of the uncommon cases, where some difficulties make the game more exciting and challenging.

> In this case we fall back into the current problem, where it is always the same boat that is summoned if the hero casts the spell several times in a row.

Except this case. I've written before, that this original rule should be in fheroes2 too and hero can summon multiple different boats for the allied heroes. (Correct me, if I'm wrong, but in heroes3 we couldn't summon different boats. So we had to summon boat, load it with other hero and then summon new one)

> (If I'm not mistaken, in the original game for some cases priority scanning is often done clockwise)

I was talking about the case, where boats are located at the same distance (the same radius).

And also, selection of summoned boat is a brilliant idea! I've never thought about this... Unfortunately, as I've written above, I'm afraid this can make the game too easy. :) But in future you need to create a suggestion issue... And if other team is not as conservative as me, they'd definitely welcome it... :D

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> Dear @LeHerosInconnu, I understand the rules you want to apply. They help to automatically summon a boat with the best conditions for a player.
> (Except summoning neutral boats first. I think it's unfair, as hero , that summoned neutral boats while having his own being built or being owned before, can rob other players. (For example, I own two boats in the beginning of the game. There's also two neutral boats somewhere on a map. I summon them from the far distant area. => now I own 4 boats. The next turn, you're trying to summon boats and it fails.)

The free boats are there as gifts for the players.
The player has the possibility to recuperate them using the "Summon Boat" spell.
It is a game strategy to do everything possible to recuperate them (if they are useful for the player's overall plan).
And it is a reward of this strategy to get them before other players.

The player has to make an effort to get the spell, it is not free.
He has to build a Mage Guild level two, and it's not certain to get it, or it is necessary to find a Shrine with the spell, or a Spell Scroll.

> But I'm so used to this "painful" spell, as it is.)) For me it's one of the uncommon cases, where some difficulties make the game more exciting and challenging.

You find that one of the flaws of the game makes it more exciting and challenging, I'm fine with that, but that's not necessarily the case for everyone.
In a strategy game, I prefer things to be well-defined with a consistent and predictable operation for a better gaming experience.

If you want, it is possible to add an option just for you that when the spell is cast the summoned boat is chosen completely randomly, so that you are happy that your plans are ruined. :D

> > In this case we fall back into the current problem, where it is always the same boat that is summoned if the hero casts the spell several times in a row.
> 
> Except this case. I've written before, that this original rule should be in fheroes2 too and hero can summon multiple different boats for the allied heroes. (Correct me, if I'm wrong, but in heroes3 we couldn't summon different boats. So we had to summon boat, load it with other hero and then summon new one)

In Heroes 3, a hero cannot summon multiple boats in all squares adjacent to that hero, as is possible in Heroes 2.
Also, in Heroes 3, depending on the level of knowledge of magic corresponding to the spell, a boat is generated automatically.
It's a different game. :)

> > (If I'm not mistaken, in the original game for some cases priority scanning is often done clockwise)
> 
> I was talking about the case, where boats are located at the same distance.

Maybe, I don't know. :)
Never done an extensive test on this subject.

> And also, selection of summoned boat is a brilliant idea! I've never thought about this... Unfortunately, as I've written above, I'm afraid this can make the game too easy. :)

It's not about making the game too easy, but to make it work in a correct and understandable way without the hassle.
In Heroes 3, this problem has been solved with the fact that the spell automatically generates a boat.
That's what I call too easy. :D

> But in future you need to create a suggestion issue... And if other team is not as conservative as me, they'd definitely welcome it... :D

Why not do it all at once, the future is now. :D

**Reply by Branikolog** [reply]
> Why not do it all at once, the future is now. :D

For now we have tons of work already.
I believe, once we start implementing new features into the game we'd spent weeks debating on each issue. For now we don't have so much spare time. :) 

> If you want, it is possible to add an option just for you that when the spell is cast the summoned boat is chosen completely randomly, so that you are happy that your plans are ruined. :D

I'm not talking about randomness here. I'm talking about simple rule: "D'you wanna summon boat? Find the nearest... And it's yours!" (surely, in the case it doesn't belong to other player)

I cannot say, original summon boat has a flaw... I think, it has some kind of feature, you have to consider. 
Here, I want to note, that such behaviour is quite predictable, while with your proposal you can forget, where was hero's boat, or what boat was built last.
Again, If we have such predictable logic, we can use "summon boat" spell for tactical methods. We can approach and summon needed boat, so enemy won't be able to capture it. 
So both proposed methods are viable.

Regarding summoning neutral boats first. You were talking about the case, when some maps have boats placed in a certain areas, so other players can reach some places or on the conrtary, leave certain areas. So summoning these boats can ruin gameplay in certatin maps. Yes, player can spam this spell and summon all that boats, but let's don't facilitate this actions. :) Let hero summon all the nearest boats: neutral or belonging to player. And only then distant neutral ones.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

Okay, let's make it work the way you wrote for now.

And I'll create a suggestion to add an option that allows the player to select the boat to summon directly from those he owns.

**Reply by Branikolog** [reply]
Okay. 
I think it would be nice if we have in future the logic, I proposed, as default, and optional enhancement, which allows to select summonning boat.

**Reply by LeHerosInconnu** [reply]
Hello @Branikolog,

Okay.
I closed my issue related to the "Summon Boat" spell https://github.com/ihhub/fheroes2/issues/2580 so that you can make your proposal.
Don't forget to indicate that it should be possible for a hero to summon a boat in all adjacent squares to the hero (https://github.com/ihhub/fheroes2/issues/2435). :)

**Comment by LeHerosInconnu** [comment]
It's okay the way it works currently.
I close the discussion.
